File: python/day12.py
from functools import reduce


instructions = [(line[0], int(line[1:])) for line in open("inputs/12.txt").readlines()]

actions = {
        'N': lambda v: 1j * v,
        'E': lambda v: v,
        'S': lambda v: -1j * v,
        'W': lambda v: -v,
        'L': lambda v: 1j**(v // 90),
        'R': lambda v: (-1j)**(v // 90),
}

manhattan = lambda pos: int(abs(pos.real) + abs(pos.imag))

move = (lambda accum, element:
        (lambda part, pos, d, action, value:
         (part, pos, d * actions[action](value))   if action in 'LR' else
         (part, pos + d * value, d)                if action == 'F' else
         (part, pos + actions[action](value),  d)  if part == 1 else
         (part, pos,  actions[action](value) + d))
        (*accum, *element))

solve = lambda part, starting_dir: manhattan(reduce(move,
                                                    instructions,
                                                    (part, 0+0j, starting_dir)
                                                   )[1])

print(solve(1,  1+0j))
print(solve(2, 10+1j))

File: nim/day12.nim
import aoc, strutils

type
  Instruction = tuple[action: char, value: int]
  Instructions = seq[Instruction]

proc parse(path: string): Instructions =
  for line in path.lines:
    result.add (line[0], parseInt(line[1..^1]))

func solve(instructions: Instructions, part: int): int =
  var dir, pos = new Point
  dir[] = if part == Part1: (1, 0) else: (10, -1)
  pos[] = (0, 0)
  var movable = if part == Part1: pos else: dir
  for instr in instructions:
    case instr.action
    of 'E': movable.x += instr.value
    of 'W': movable.x -= instr.value
    of 'S': movable.y += instr.value
    of 'N': movable.y -= instr.value
    of 'F':
      pos.x += dir.x * instr.value
      pos.y += dir.y * instr.value
    of 'L':
      for _ in 1 .. (instr.value div 90):
        dir[] = (dir.y, -dir.x)
    of 'R':
      for _ in 1 .. (instr.value div 90):
        dir[] = (-dir.y, dir.x)
    else: discard
  result = manhattan pos[]


let input = parse "inputs/12.txt"
echo input.solve Part1
echo input.solve Part2
